internal/testrunner/runners/pipeline/coverage.go

```diff
@@ -0,0 +1,135 @@
+		return nil, errors.Wrap(err, "locating data_stream root failed")
+	}
+	if !found {
+		return nil, errors.New("data stream root not found")
+	}
+
+	// Use the Node Stats API to get stats for all installed pipelines.
+	// These stats contain hit counts for all main processors in a pipeline.
+	stats, err := ingest.GetPipelineStats(options.API, pipelines)
+	if err != nil {
+		return nil, errors.Wrap(err, "error fetching pipeline stats for code coverage calculations")
+	}
+
+	// Construct the Cobertura report.
+	pkg := &testrunner.CoberturaPackage{
+		Name: options.TestFolder.Package + "." + options.TestFolder.DataStream,
+	}
+
+	coverage := &testrunner.CoberturaCoverage{
+		Sources: []*testrunner.CoberturaSource{
+			{
+				Path: packagePath,
+			},
+		},
+		Packages:  []*testrunner.CoberturaPackage{pkg},
+		Timestamp: time.Now().UnixNano(),
+	}
+
+	// Calculate coverage for each pipeline
+	for _, pipeline := range pipelines {
+		covered, class, err := coverageForSinglePipeline(pipeline, stats, packagePath, dataStreamPath)
+		if err != nil {
+			return nil, errors.Wrapf(err, "error calculating coverage for pipeline '%s'", pipeline.Filename())
+		}
+		pkg.Classes = append(pkg.Classes, class)
+		coverage.LinesValid += int64(len(class.Methods))
+		coverage.LinesCovered += covered
+	}
+	return coverage, nil
+}
+
+func coverageForSinglePipeline(pipeline ingest.Pipeline, stats ingest.PipelineStatsMap, packagePath, dataStreamPath string) (linesCovered int64, class *testrunner.CoberturaClass, err error) {
+	// Load the list of main processors from the pipeline source code, annotated with line numbers.
+	src, err := pipeline.Processors()
+	if err != nil {
+		return 0, nil, err
+	}
+
+	pstats, found := stats[pipeline.Name]
+	if !found {
+		return 0, nil, errors.Errorf("pipeline '%s' not installed in Elasticsearch", pipeline.Name)
+	}
+
+	// Ensure there is no inconsistency in the list of processors in stats vs obtained from source.
+	if len(src) != len(pstats.Processors) {
+		return 0, nil, errors.Errorf("processor count mismatch for %s (src:%d stats:%d)", pipeline.Filename(), len(src), len(pstats.Processors))
+	}
+	for idx, st := range pstats.Processors {
+		// Check that we have the expected type of processor, except for `compound` processors.
+		// Elasticsearch will return a `compound` processor in the case of `foreach` and
+		// any processor that defines `on_failure` processors.
+		if st.Type != "compound" && st.Type != src[idx].Type {
+			return 0, nil, errors.Errorf("processor type mismatch for %s processor %d (src:%s stats:%s)", pipeline.Filename(), idx, src[idx].Type, st.Type)
+		}
+	}
+
+	// Tests install pipelines as `filename-<nonce>` (without original extension).
+	// Use the filename part for the report.
+	pipelineName := pipeline.Name
+	if nameEnd := strings.LastIndexByte(pipelineName, '-'); nameEnd != -1 {
+		pipelineName = pipelineName[:nameEnd]
+	}
+
+	// File path has to be relative to the packagePath added to the cobertura Sources list
```

Comment:
Let's discuss the safe deployment of this feature.

If we merge this PR and release a new elastic-package, will it break the CI tests for Elastic Integrations? If it breaks, then we need to go the other way round and adjust the Jenkinsfile first and make sure that pipeline sources are in the right directory (relative to the runner?).

Reply:
There's no reason to believe it will break any tests, I've tested it multiple times with all the existing integrations.

I think you're misunderstanding the CI issues that will need to be addressed. I explained them in this comment: https://github.com/elastic/elastic-package/pull/585#issuecomment-970186919

Those issues will not cause a test failure. Just the coverage report displayed in CI will not show annotated source code.

Reply:
Yes, that's what I mean. Let me share my concerns, Jenkins uses [coverageReport](https://github.com/elastic/integrations/blob/main/.ci/Jenkinsfile#L280) function to collect the coverage stats. I don't know what will happen if that function files (due to missing pipeline sources). Will it not show the coverage split by source code or fail the entire CI pipeline?

Reply:
It is not causing a CI pipeline failure in this PR, just the source code is reported missing. See:

https://beats-ci.elastic.co/job/Ingest-manager/job/elastic-package/job/PR-704/4/cobertura/apache_access/apache_data_stream_access_elasticsearch_ingest_pipeline_default_yml/

Reply:
Understood, thanks for clarifying this.